### DISCUSSION TITLE
Max connections.

### DISCUSSION THREAD

**Discussion by n-hor** [qruto/laravel-wave#58] [General]
If I understand correctly, the maximum number of sse connections = the maximum number of PHP-fpm workers,
So if I have 20 php-fpm workers and 20 connected users, will my application stop responding?

**Comment by slavarazum** [comment]
The sort answer - yes, so you should adjust configuration values with your specific requirements. SSE connections are processed as standard http requests by your server.

**Reply by n-hor** [reply]
I don’t understand the sense of this package. without non-blocking i/o approach it will block the app. Octane doesn't support coroutines and the app will be unavailable if all workers handle SSE.

**Reply by slavarazum** [reply]
> The essence of the package is to eliminate the need to setup and maintain a WebSockets server separately from the web server. 

With Wave, you only need a standard web server that processes HTTP requests to get all of the broadcasting(real-time) purposes.

If your php-fpm worker is configured for a maximum of 20 connections and 20+ requests are sent to your server concurrently, then the application will also be blocked. This is a standard question of scaling a web server for incoming traffic, whether you are using SSE or not.

**Reply by n-hor** [reply]
The task of the php worker, or the octane worker, is to process the request as quickly as possible. This package creates a potential denial of service vulnerability for users who don't understand how it works under the hood. I think that it is still worth warning users in the package description that they should not use it in production, or at least add a description that if you have 4 fpm/octane workers and 4 users subscribe to the event, the application will unavailable.

It's not about server scaling, 20 php-fpm workers that can handle 200-300RPS on a simple server with your package will be stuck with 20 online users.

**Reply by IfnotFr** [reply]
I am jumping on this for my real use case : I am thinking about switching to SSE instead of websocket for my app [check-position.com](https://check-position.com/) which has several hundred of online users regulary.

I have 150 php workers on my web pool, so using this package will down my complete setup. This package is dangerous as is and not suitable for large apps, you should explain on the home page the resources limitations compared to a regular event driven system.

**Reply by slavarazum** [reply]
Currently I'm working on extended production server setup and scaling documentation which will describe more clearly that you can use it in production as well as for large scale apps.

However, I think it should be clear that when transferring the responsibility for maintaining real-time connections from WebSockets to Web Server, the load on it will obviously increase.
I also want to note that 200-300RPS != 200-300 concurrent connections.

Feel free to open a Pull Request with any suggestion you think would improve the Wave experience.

**Reply by n-hor** [reply]
200-300 requests per second which workers will process in order. Your package COMPLETELY blocks the worker for 1 connection. 

All you have to do is open your demo in several tabs and it **DIES**. How can this approach even be compared with websockets? Why not just warn users about this? Why are you trying to talk about scalability and settings if the implementation approach is completely incorrect?

Think about the unfortunate users who installed your package and don’t understand why their application stopped working.

**Comment by IfnotFr** [comment]
Be more respecful @seqond-soft , the idea is great, but dangerous for production that's all. It only should me warned on the documentation. I don't know how it is built but moving it to octane / swoole with event implementation should be by default and if the user still want to use it with fpm it should warn him for potential performance issue (it may be not a problem for a small personnal app).

**Comment by slavarazum** [comment]
I understand your point that it can be frustrating without extra knowledge or instructions.
Just to notice that the package is currently under version 0.8.* and with the right server setup for your traffic needs, it will not bring down the server.
Currently, version 1.0 is being prepared, which will include expanded documentation.

**Comment by abrhamsw** [comment]
> the maximum number of PHP-fpm workers

Thank you for your information 
To be clear, I built a chat app that is used by 1000 clients that is deployed on shared hosting / Cpanel 
as you may know, most shared hosting is 10 to 30 php-fpm workers.

So based on your information I can't use this package for this use case ?

**Reply by n-hor** [reply]
No. With this package one php-fpm worker can serve one online user.

**Reply by abrhamsw** [reply]
Thank you for your replay 

do you know any alternatives package that solve this problem ?

**Comment by slavarazum** [comment]
In this case you have 2 options:

1. Increase amount of available workers in config.
2. Use Octane to handle multiple connections with a single worker avoiding php-fpm altogether.

**Reply by n-hor** [reply]
In the current implementation, octane doesn't support coroutines and connection pools, so one SSE user, just like with fpm, will permanently take one worker.

**Reply by gw8484** [reply]
Just figured out this problem after a few days of trying various toy Laravel SSE examples online and AI hallucinated examples based on non-existent coroutine support in Laravel.  But to clarify, the root problem is not actually in Octane but Laravel itself which is not capable of working with coroutines and there is no plan to add the support.  So, coroutine capable PHP servers (e.g. Swoole, etc.) are effectively crippled under Laravel.

This problem is not limited to Laravel Wave but all SSE libraries/apps using the standard single Laravel (Octane or not) server to handle SSE connections.  Please correct me if I am wrong but it seems that Laravel Wave cannot solve this problem without having a separate dedicated multi-threaded/coroutine based SSE servers that can handle large number of concurrent SSE connections per worker.  It's obvious now why Laravel Reverb uses a separate dedicated server.

I wish Laravel and Laravel Wave had clear warnings about this problem.
